NetworkSpec: support user supplied infrastructure

vincepri opened the pull request:
Signed-off-by: Vince Prignano <vincepri@vmware.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
This PR introduced `NetworkSpec` in the Cluster configuration object. Users can supply new configuration options for managed VPCs/Subnets or supply a pre-existing infrastructure by ids.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #267 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

vincepri commented on pkg/cloud/aws/services/ec2/vpc.go:
```diff
@@ -88,29 +107,46 @@ func (s *Service) createVPC() (*v1alpha1.VPC, error) {
+	}
+
+	if err := tags.Apply(tagApply); err != nil {
+		return nil, err
```
That's true, the behavior there is the same as before.

detiber replied:
eep, that's not good. I thought we had fixed it up a while ago so that it would warn only on tagging failures :(

vincepri replied:
Filed issue #539

vincepri commented on pkg/apis/awsprovider/v1alpha1/types.go:
```diff
@@ -104,16 +105,21 @@ type Network struct {
 }
 
+// IsManaged returns true if the VPC is managed by cluster api.
+func (v *VPC) IsManaged() bool {
```
Replied below.

vincepri replied:
I thought about it and went back and forth for a while on naming. `IsManaged` is the most explicit given that it matches the tag `sigs.k8s.io/cluster-api-provider-aws/managed` = `true`. But I have no strong opinions here, if that's a general preference.

vincepri commented on pkg/cloud/aws/services/ec2/gateways.go:
```diff
@@ -35,6 +35,10 @@ func (s *Service) reconcileInternetGateways() error {
 
 	igs, err := s.describeVpcInternetGateways()
 	if awserrors.IsNotFound(err) {
+		if !s.scope.VPC().IsManaged() {
```
That's where I think we need user stories, I've been in companies where teams would at least have read permissions on what they were supposed to use. I propose to assume we can at least read to start, if there are user stories later on I'm happy to revisit.

vincepri commented on pkg/cloud/aws/services/ec2/gateways.go:
```diff
@@ -45,6 +49,11 @@ func (s *Service) reconcileInternetGateways() error {
 	}
 
 	gateway := igs[0]
```
For the first iteration I think we should say the network is either managed or unmanaged.

That means when you provide a VPC ID, you need to provide everything else too. The contract in an unmanaged scenario is that we'd read information about the network configuration and perform no create/modify/delete actions on the components.

vincepri commented:
/retest

vincepri commented:
/test pull-cluster-api-provider-aws-bazel-e2e

vincepri commented:
/hold cancel

vincepri commented:
@chuckha @detiber ptal, updated with all the comments

vincepri commented on pkg/apis/awsprovider/v1alpha1/awsclusterproviderconfig_types.go:
```diff
@@ -41,10 +41,62 @@ type AWSClusterProviderSpec struct {
+}
+
+// SubnetSpec configures an AWS Subnet.
+type SubnetSpec struct {
```
This should be resolved in the latest push, marking as outdated

vincepri commented:
@detiber I'll open an issue for `clusterawsadm` support and update the documentation.

@randomvariable That's a good point. For the first iteration, supplied infrastructure should be complete, meaning that the actuator won't be adding any primary resource, but it'll create security groups, at least for now. I actually need to make sure both the NAT and Internet gateways return early and make no modifications (or describe calls), looking at the PR seems the change got lost in one of the rebase I made.

vincepri commented:
@randomvariable ptal

randomvariable commented:
Can we add some docs on which use cases are supported. Ideally, can we link it back to the enumerated state space of the original [feature model](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/proposal/features.md) ?

IF we don't do this now, it will fall on whoever does #519, who will have to reverse engineer intent.

vincepri commented:
I'm writing docs next, need to validate the workflow first. Could you take a look at the early returns in the resources that you mentioned in your last comment?

randomvariable commented:
Yup, looks ok. Can't see any calls to EC2 services that shouldn't happen in the IsManaged() scenario.

vincepri commented:
@chuckha @detiber @randomvariable I've updated the PR with some docs and limitations + manifest script changes. Also fixed some issues that I encountered when having the same name for ELBs, instances and subnets.

vincepri commented on pkg/cloud/aws/tags/types.go:
```diff
@@ -26,6 +29,24 @@ func (m Map) Equals(other Map) bool {
 }
 
+// HasOwned returns true if the tags contains a tag that marks the resource as owned by the cluster.
+func (m Map) HasOwned(cluster string) bool {
```
That's gonna be an interesting use case, when the VPC is shared between clusters, but that's probably out of scope for v1alpha1.

detiber replied:
I'm not sure we should consider it out of scope. In the case of separation of concerns users might very well choose to re-use the same VPC to avoid having to request a VPC from another team for each cluster they create.

chuckha commented on pkg/cloud/aws/services/ec2/subnets.go:
```diff
@@ -113,24 +117,31 @@ LoopExisting:
-		if subnet.ID != "" {
-			continue
-		}
+	if !s.scope.VPC().IsProvided() {
```
[nit, non-blocking] The rest of the code seems to check provided first then return, why did this flip logic? I'd have expected to see

```go
if s.scope.VPC().IsProvided() {
    klog.V(2).Infof("Provided subnets: %s", subnets)
    return nil
}
```

vincepri replied:
Here we want to describe the subnets, the subnet-id is optional from a user perspective, while the VPC is required. The other components return early because they don't need to be described or validated at this point.

vincepri commented:
Potential users are customers that have VPCs and network managed by a network team, folks that want to keep using the AWS quickstart or people that manage network with other tools like Terraform.

sfzylad commented on docs/getting-started.md:
````diff
@@ -166,6 +166,24 @@ customization now is the time to do it. Take a look at
+By default the cluster name is set `test1` when generating manifests, `CLUSTER_NAME` environment variable is used in `make manifests` to customize the cluster name.
+
+```bash
+export CLUSTER_NAME="<pony-unicorns>"
````
This variable seems not to work. I set it up and `make manifests{,-dev}` still seems to generate manifests for cluster named `test1`. Please let me know if I'm doing something wrong.

vincepri commented:
/test pull-cluster-api-provider-aws-bazel-e2e

vincepri commented:
/retest

detiber commented on cmd/clusterctl/examples/aws/cluster-network-spec.yaml.template:
```diff
@@ -0,0 +1,20 @@
+      sshKeyName: "${SSH_KEY_NAME}"
+      networkSpec:
+        vpc:
+          id: "${VPC_ID}"
```
Instead of having a completely separate (and mostly duplicated) template here, would it make sense to use kustomize for adding the networkSpec to the existing template for this purpose instead?

vincepri replied:
I thought about it, but Kustomize isn't powerful enough to allow for user-input or get environment flags. I've seen other projects having the same issue and using `sed` to overcome these problems. I think in the future we should template kustomize files instead using maybe something like go templates, or similar environment substitution tool.

detiber replied:
The issue with trying to override env vars at runtime is a known issue with kustomize (and one that they don't currently plan on supporting), we work around a similar issue to this in how we manage the aws_manager_image_patch.yaml file.

vincepri replied:
Opened https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/549 to track :)

detiber commented:
/lgtm
